### ISSUE TITLE
automatic self-healing (bitrot and unrer-replication correction) [feature]

### ISSUE THREAD

**Issue by onlyjob** [chrislusf/seaweedfs#1519]
Ability to self-repair under-replicated data (due to missing volumes or data corruption) is one of the most important features for production readiness. Of all storage systems, _MooseFS_ implemented the best approach for detection and automatic repair of problems.

On all chunkservers (alternative to SeaweedFS's volume servers) a background scrub is running on each HDD. Speed of background read is configurable. Once an error is detected (e.g. read error, CRC error, etc.) a chunk is marked as "undergoal" (under-replicated). Replication is always running in the background to ensure desirable number of replicas so undergoal chunks are detected and replicated within minutes, effectively correcting all types of errors in data.  
_There are no periodic tasks, cron jobs or anything that storage administrator have to do (manually) to check and ensure correct replication level._  
MooseFS's chunkservers notify master about graceful shutdown. When chunkserver is stopped, master do not trigger replication for some time, giving chunkserver enough time to come back after maintenance. In case of sudden disappearance of chunkserver, there is a configurable delay before replication begins to re-create missing chunks elsewhere. Another delay is to postpone replication of data for some time after start of _master_ to allow all chunkserver to come online and register with master. There is also an option to put chunkserver insto maintenance mode on master to postpone replication of its data.  
MooseFS's chunkservers are smart enough to mark a particular HDD as read-only if number of errors are above certain threshold. In such case, one failed HDD is eventually excluded if it becomes unreliable.
(MooseFS's chunkservers can even (re-)balance data between local HDDs for even utilisation.)

Basically periodic data scrubbing should be performed automatically when volumes are idle.
`volume.fix.replication` should run automatically too, when needed (e.g. when there is something to replicate), on stable cluster when all volumes are available or when a volume server is missing for too long.  
Without those measures data is always at risk.

Please consider implementing automatic checks and corrections. Thanks.

**Comment by chrislusf** [comment]
Thanks! Some of these ideas will get into the system.

**Comment by timolow** [comment]
I would recommend taking a look at how Openstack Swift deals with bitrot and CRC issues as a good starting off point.

**Comment by rpodgorny** [comment]
any news on this one? i'm considering migrating from moosefs but a missing auto-heal functionality sounds like a recipe for serious data loss... :-(

**Comment by chrislusf** [comment]
There are multiple shell commands added:
```
$ weed shell
master: localhost:9333 filers: [192.168.2.10:8888]
> help
Type:	"help <command>" for help on <command>. Most commands support "<command> -h" also for options.
  cluster.ps                    	# check current cluster process status
  collection.delete             	# delete specified collection
  collection.list               	# list all collections
  ec.balance                    	# balance all ec shards among all racks and volume servers
  ec.decode                     	# decode a erasure coded volume into a normal volume
  ec.encode                     	# apply erasure coding to a volume
  ec.rebuild                    	# find and rebuild missing ec shards among volume servers
  fs.cat                        	# stream the file content on to the screen
  fs.cd                         	# change directory to a directory /path/to/dir
  fs.configure                  	# configure and apply storage options for each location
  fs.du                         	# show disk usage
  fs.ls                         	# list all files under a directory
  fs.meta.cat                   	# print out the meta data content for a file or directory
  fs.meta.load                  	# load saved filer meta data to restore the directory and file structure
  fs.meta.notify                	# recursively send directory and file meta data to notifiction message queue
  fs.meta.save                  	# save all directory and file meta data to a local file for metadata backup.
  fs.mkdir                      	# create a directory
  fs.mv                         	# move or rename a file or a folder
  fs.pwd                        	# print out current directory
  fs.rm                         	# remove file and directory entries
  fs.tree                       	# recursively list all files under a directory
  lock                          	# lock in order to exclusively manage the cluster
  remote.cache                  	# cache the file content for mounted directories or files
  remote.configure              	# remote storage configuration
  remote.meta.sync              	# synchronize the local file meta data with the remote file metadata
  remote.mount                  	# mount remote storage and pull its metadata
  remote.mount.buckets          	# mount all buckets in remote storage and pull its metadata
  remote.uncache                	# keep the metadata but remote cache the file content for mounted directories or files
  remote.unmount                	# unmount remote storage
  s3.bucket.create              	# create a bucket with a given name
  s3.bucket.delete              	# delete a bucket by a given name
  s3.bucket.list                	# list all buckets
  s3.clean.uploads              	# clean up stale multipart uploads
  s3.configure                  	# configure and apply s3 options for each bucket
  unlock                        	# unlock the cluster-wide lock
  volume.balance                	# balance all volumes among volume servers
  volume.check.disk             	# check all replicated volumes to find and fix inconsistencies
  volume.configure.replication  	# change volume replication value
  volume.copy                   	# copy a volume from one volume server to another volume server
  volume.delete                 	# delete a live volume from one volume server
  volume.deleteEmpty            	# delete empty volumes from all volume servers
  volume.fix.replication        	# add or remove replicas to volumes that are missing replicas or over-replicated
  volume.fsck                   	# check all volumes to find entries not used by the filer
  volume.list                   	# list all volumes
  volume.mark                   	# Mark volume writable or readonly from one volume server
  volume.mount                  	# mount a volume from one volume server
  volume.move                   	# move a live volume from one volume server to another volume server
  volume.tier.download          	# download the dat file of a volume from a remote tier
  volume.tier.move              	# change a volume from one disk type to another
  volume.tier.upload            	# upload the dat file of a volume to a remote tier
  volume.unmount                	# unmount a volume from one volume server
  volume.vacuum                 	# compact volumes if deleted entries are more than the limit
  volumeServer.evacuate         	# move out all data on a volume server
  volumeServer.leave            	# stop a volume server from sending heartbeats to the master
>
```

In master.toml, these operations are run every 17 minutes to "self-heal".
```
scripts = """
  lock
  ec.encode -fullPercent=95 -quietFor=1h
  ec.rebuild -force
  ec.balance -force
  volume.deleteEmpty -quietFor=24h -force
  volume.balance -force
  volume.fix.replication
  unlock
"""
sleep_minutes = 17          # sleep minutes between each script execution

```

**Comment by rpodgorny** [comment]
thanks for the info! ...is the 17-minute script enabled by default (compiled-in) or am i supposed to add it manually?

**Comment by chrislusf** [comment]
it is in the `master.toml` generated in `weed scaffold -config=master`. You can tweak it.

**Comment by onlyjob** [comment]
IMHO system should repair errors _out-of-the-box_, by default. That default can be a subject of tweaking (not the absence of default). Why is repair meant to be _do-it-yourself_??

**Comment by chrislusf** [comment]
The repair is default, and can be overwritten by the settings in
master.toml file.
